src/main/scala/com.snowplowanalytics.snowplow.badrows/Payload.scala

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright (c) 2018-2019 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+
+package com.snowplowanalytics.snowplow.badrows
+
+import cats.syntax.functor._
+import io.circe.{Decoder, Encoder}
+import io.circe.generic.auto._
+import io.circe.syntax._
+
+sealed trait Payload
+object Payload {
+  implicit val payloadEncoder: Encoder[Payload] = Encoder.instance {
+    case p: RawPayload => p.asJson
+    case p: CollectorPayload => p.asJson
+    case p: PartiallyEnrichedEvent => p.asJson
+  }
+  implicit val payloadDecoder: Decoder[Payload] = List[Decoder[Payload]](
+    Decoder[RawPayload].widen,
+    Decoder[CollectorPayload].widen,
+    Decoder[PartiallyEnrichedEvent].widen
+  ).reduceLeft(_ or _)
+}
+final case class RawPayload(line: String) extends Payload
+
+final case class NVP(name: String, value: Option[String])
+final case class CollectorPayload(
+  vendor: String,
+  version: String,
+  querystring: List[NVP], // Could be empty in future trackers
+  contentType: Option[String], // Not always set
+  body: Option[String], // Not set for GETs
+  collector: String,
+  encoding: String,
+  hostname: Option[String],
+  timestamp: Option[String], // Must have a timestamp
+  ipAddress: Option[String],
+  useragent: Option[String],
+  refererUri: Option[String],
+  headers: List[String], // Could be empty
+  networkUserId: Option[String] // User ID generated by collector-set third-party cookie
+) extends Payload
+
+final case class PartiallyEnrichedEvent(
```

Comment:
I thought at this time we already can assume `UUID`/`Instant`?

Reply:
no it's already strings